src/tensor_ops/choose/mod.rs

```diff
@@ -0,0 +1,147 @@
+mod cpu_kernel;
+
+#[cfg(feature = "cuda")]
+mod cuda_kernel;
+
+use crate::{
+    gradients::{Merge, Tape},
+    prelude::{DeviceStorage, HasErr, PutTape, SplitTape, Tensor},
+    shapes::{Dtype, Shape},
+};
+
+pub trait ChooseKernel<E: Dtype>: DeviceStorage {
+    fn forward<S: Shape>(
+        &self,
+        cond: &Self::Storage<S, bool>,
+        lhs: &Self::Storage<S, E>,
+        rhs: &Self::Storage<S, E>,
+    ) -> Result<Self::Storage<S, E>, Self::Err>;
+
+    fn backward<S: Shape>(
+        &self,
+        cond: &Self::Storage<S, bool>,
+        grad_lhs: &mut Self::Storage<S, E>,
+        grad_rhs: &mut Self::Storage<S, E>,
+        grad_out: &Self::Storage<S, E>,
+    ) -> Result<(), Self::Err>;
+}
+
+/// Choose values from two tensors using a boolean mask. Equivalent to `torch.where` from pytorch.
+pub trait ChooseFrom<Lhs, Rhs>: HasErr {
+    type Output;
+
+    /// Construct a new tensor, where the output tensor contains the elements of lhs where self is
+    /// true, and rhs where self is false.
+    fn choose(self, lhs: Lhs, rhs: Rhs) -> Self::Output {
+        self.try_choose(lhs, rhs).unwrap()
+    }
+
+    /// Fallible version of choose
+    fn try_choose(self, lhs: Lhs, rhs: Rhs) -> Result<Self::Output, Self::Err>;
+}
+
+impl<
+        S: Shape,
+        E: Dtype,
+        D: ChooseKernel<E>,
+        LhsTape: Tape<D> + Merge<RhsTape>,
+        RhsTape: Tape<D>,
+    > ChooseFrom<Tensor<S, E, D, LhsTape>, Tensor<S, E, D, RhsTape>> for Tensor<S, bool, D>
+{
+    type Output = Tensor<S, E, D, LhsTape>;
+
+    fn try_choose(
+        self,
+        lhs: Tensor<S, E, D, LhsTape>,
+        rhs: Tensor<S, E, D, RhsTape>,
+    ) -> Result<Self::Output, Self::Err> {
+        let (lhs, tape) = lhs.split_tape();
```

Comment:
Similar to other comment, but need to add an assert for shape equality for both pairs:
```rust
assert_eq!(self.shape(), lhs.shape());
assert_eq!(self.shape(), rhs.shape());
```